### ISSUE TITLE
SPARQL query on SHACL shape graph returns no results

### ISSUE THREAD

**Issue by jeenbroekstra** [eclipse/rdf4j#2669]
Although it is possible to access the statements in the SHACL Shape graph by means of a `getStatements` request:

    conn.getStatements(null, null, null, RDF4J.SHACL_SHAPE_GRAPH);

The equivalent SPARQL query yields no results:

    SELECT * 
    FROM <http://rdf4j.org/schema/rdf4j#SHACLShapeGraph> 
    WHERE { ?s ?p ?o }

Underlying cause is that internally, the shape graph is kept in a separate repository, and while getStatements calls on the shape graph get intercepted and redirected, the same does not happen for SPARQL queries with the "magic graph name" in their Dataset. 

This is problematic for use cases where the shape graph is not just used for transactional validation, but also for things like driving a UI (and therefore needs to be accessible to query). 

A possible workaround is to create a copy of the shapes graph in a different repository or in a "normal" named graph in the same repository.

**Comment by hmottestad** [comment]
This is a known limitation yes. Would be nice to have this, but the effort to support it would be immense. I think the more viable approach would be allowing users to store shapes outside of the special graph. 

Options
- user has to specify a graph (easier to keep track of changes)
- user can use any or no graph (more expensive if we have to check the namespace of all added statements or if we have to retrieve all the shapes every validation)

**Comment by jeenbroekstra** [comment]
> This is a known limitation yes. Would be nice to have this, but the effort to support it would be immense. 

Can you elaborate on why you think that would be so difficult? Internally, the shapes are kept in a repo, so as far as I can tell the only thing required is to intercept incoming SPARQL queries, inspect their Dataset definition, and if it uses the "magic" named graph, just redirect the query to the internal shapes repo instead of the base sail.

**Comment by jeenbroekstra** [comment]
If I can find the time I may do a small PoC of the above idea. If it turns out to be more complicated than I anticipate we can step back and reasses

**Comment by amivanoff** [comment]
We did a little prototype with NotifyingSailConnectionWrapper and (Public Morozov antipattern with org.apache.commons.lang3.reflect.FieldUtils due to private fields in ShaclSailConnection) and doing some tests.

`public class SparqledShaclSailConnection extends NotifyingSailConnectionWrapper {
	public SparqledShaclSailConnection(NotifyingSailConnection connection) {
		super(connection);
	}

	@Override
	public CloseableIteration<? extends BindingSet, QueryEvaluationException> evaluate(TupleExpr tupleExpr, Dataset dataset, BindingSet bindings, boolean includeInferred) throws SailException {
		if (dataset == null) {
			return super.evaluate(tupleExpr, dataset, bindings, includeInferred);
		} else {
			List<IRI> graphs = new ArrayList<IRI>();
			graphs.addAll(dataset.getDefaultGraphs());
			graphs.addAll(dataset.getNamedGraphs());
			
			if (graphs.contains(RDF4J.SHACL_SHAPE_GRAPH)) {
				SailRepositoryConnection shapesRepoConnection;
				try {
					shapesRepoConnection = (SailRepositoryConnection) FieldUtils.readField(this.getWrappedConnection(), "shapesRepoConnection", true);
					return shapesRepoConnection.getSailConnection().evaluate(tupleExpr, null, bindings, includeInferred);
				} catch (IllegalAccessException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
					return super.evaluate(tupleExpr, dataset, bindings, includeInferred);
				}
			} else {
				return super.evaluate(tupleExpr, dataset, bindings, includeInferred);
			}
		}
	}
}`

It seems, Sparql Select and Insert are both working. But incorrect data from rejected transaction sneaking into repository somehow.

**Comment by amivanoff** [comment]
For the tests I am using the latest RDF4J 3.5.0M1.

**Comment by hmottestad** [comment]
Ok. I was assuming you were going to do some sort of federated query or query rewriting. Also thought you might want to support variables in the graph clause. 

One thing to keep in mind when handling this is to trigger the revalidation and updating the cache.

**Comment by amivanoff** [comment]
In this case it's a primitive graph-name based query dispatching between an internal shapesRepoConnection from ShaclSail and a main repository connection. Like the approach in a getStatements() method.
It's not the "full-scale federation" like FedX with results aggregation, etc...

**Comment by amivanoff** [comment]
I'm trying to realize exactly the "**shape-driven web UI**" use-case with RDF4J. It includes two issues with shapes:
* "shape querying" to create UI from schema (react-jsonschema-form like)
* and "runtime shape modification" (run-time modifiable data model with run-time UI adaptation).

Our initial prototypes works with SPARQL Select and SPARQL Update while we've used shapes as RDFS data in default graph (without validation) with the Sail stack contains SchemaCachingRDFSInferencer only.

After the DirectTypeHierarchyInferencer sail addition into the stack it breaks with a deadlock on initial shapes upload into the default graph. So, I decided to reevaluate the state of ShaclSail because it treats shapes differently in a separated graph.

I've done some simple concurrency data load benchmarks today trying to load several turtle files with vocabs (12), shapes (4) and data (6) into a repository. All tests are without SPARQL at all. All files are correct.

Repository consists of several sails (from top to bottom):
* SparqledShaclSail (with SparqledShaclSailConnection, mentioned before),
* DirectTypeHierarchyInferencer,
* SchemaCachingRDFSInferencer
* and NativeStore.

My results supports recommendations from the Documentation: "use bulk-load operations for shapes" ))
* Sequential files load (one-by-one) in one thread works well with transactions and without them.
* Parallel (one-by-one) loading of vocabs and data files with bulk load a model with all of shapes data also works.
* Parallel load of all files (including one-by-one shape files load) in 22 threads getting stuck in a deadlock. With transactions and without them, no difference. Also get some messages like this:
`[pool-3-thread-19] INFO org.eclipse.rdf4j.common.concurrent.locks.LockManager - Thread pool-3-thread-19 is waiting on an active Write lock acquired in pool-3-thread-13`.

So, my point, that there are 2 issues from the "shape-driven web UI" perspective:
* Select shapes with SPARQL Select (it is almost ready, if @hmottestad will not object) -- I think it is what @jeenbroekstra means in this issue initially
* Modify shapes with SPARQL Update (much more complex issue, or maybe we could workaround it with revalidation triggering) -- It is what @hmottestad means in his answers

**Comment by hmottestad** [comment]
Try to set: isShapeRefreshNeeded

**Comment by hmottestad** [comment]
Also. I've done very little testing of using a reasoner sail together with the SHACL sail. Could you post some example code that reproduces the deadlock you mention?

**Comment by amivanoff** [comment]
@hmottestad I've done minimal test in a forked rdf4j repo as a subclass of the MultithreadedNativeStoreTest class.
See https://github.com/agentlab/rdf4j/commit/205b6cfce7a3db91d6e72d203814e067a70a5272

I've tried it with RDFS+Native sail stask and DT+RDFS+Native sail stask. It deadlocks only on 2 tests:
* testLotsOfValidationFailuresSerializableValidation
* testDataAndShapes

**Comment by hmottestad** [comment]
I've finally gotten around to testing with the rdfs reasoner and you are right that it does deadlock. Since I wrote both the SHACL and the rdfs code I'm inclined to believe it must be a bug with the native store :P Might try to find out what is going on though, but to be honest these kind of errors can take many hours to track down and it could end up just being related to the inference serialization bug we already know of and can't fix without a major rewrite of a fair chunk of the general sail code. I'll see if I can't make an issue for this together with a working example.

**Comment by hmottestad** [comment]
On another note. This issue was originally about using SPARQL to access the shapes. 

With the new AST in place I propose the following solution:

 - user-configurable named graph
 - monitor said graph for changes (notifying sail)
 - clear shapes cache and copy over the contents of said named graph if there are changes
